### ISSUE TITLE
Understanding the logic of custom transformers

### ISSUE THREAD

**Issue by nejatb** [jpmml/sklearn2pmml-plugin#1]
Hi Villu,

I am trying to implement a custom transformer for PMMLpipeline and have been following your example plugin. On the python side, I have implemented a class called Hasher that is of the TransformerMixin type. In the training code I have:      

               mapper = DataFrameMapper([

                    ("feature1", TfidfVectorizer(stop_words='english',
                                    norm=None,
                                    tokenizer=Splitter(),
                                    ngram_range=(1, 1)
                                    )),
                    ("feature2", TfidfVectorizer(stop_words='english',
                                 norm=None,
                                 tokenizer=Splitter(),
                                 ngram_range=(1, 1),
                                 max_df=0.9,
                                 min_df=3
                                 )),
                    ("feature3",Hasher(n_features=1,input_type="string", non_negative=True))
                    ],df_out=True)

In the java side, I implement a class that extends Transformer and I override encodeFeatures. I followed your example as well as TfidfVectorizer in Jpmml-sklearn library. I was hoping to have the value of feature3 in my model, something like: 

     <DerivedField name="hasher(feature3)" optype="continuous" dataType="double">
			<Apply function="hasher">
				<FieldRef field="feature3"/>
				<Constant dataType="string">abc</Constant>
                            <Constant dataType="double">2.193843</Constant>
			</Apply>
	</DerivedField>

where "abc" is the actual feature value and 2.193843 is the hashed value. But in encodeFeatures, when I look into the list of Features, each Feature only has a name and that happens to be feature3. I tried following TfidfVectorizer and keep a vocabulary (dictionary) of all the seen values that I named "FHvocabulary", but when I try to get the vocabulary using 

      public Map<String, ?> getVocabulary(){
              return (Map)get("FHvocabulary");
         }

I get the error that :

     java.lang.IllegalArgumentException: Expected 64 element(s), got 63 element(s)
	     at org.jpmml.sklearn.ClassDictUtil.checkSize(ClassDictUtil.java:130)
	     at sklearn.pipeline.Pipeline.encodeFeatures(Pipeline.java:76)
	     at sklearn2pmml.PMMLPipeline.encodePMML(PMMLPipeline.java:119)
	     at org.jpmml.sklearn.Main.run(Main.java:144)
	     at org.jpmml.sklearn.Main.main(Main.java:93)

Could you point me to the right direction of how I can pass the feature values and their hashed values to the java code?
Thank you so much.

**Comment by vruusmann** [comment]
> Hasher(n_features=1,input_type="string", non_negative=True)

What's happening inside `Hasher.transform(X)`? Can you share its Python code?

Based on the above usage, it seems to me that this method accepts a column of strings, and produces a column of numbers.

The corresponding Java code would look like this:
```Java
public class Hasher extends Transformer {

  @Override
  public List<Feature> encodeFeatures(List<Feature> features, SkLearnEncoder encoder){
    ClassDictUtil.checkSize(features, 1); // Consume 1-D input
    Feature feature = features.get(0);
    DerivedField derivedField = createHasherImplementationInPMML(feature);
    return Collections.singletonList(new ContinuousFeature(encoder, derivedField)); // Produce 1-D output
  }
}
```

**Comment by vruusmann** [comment]
> java.lang.IllegalArgumentException: Expected 64 element(s), got 63 element(s)
>   at org.jpmml.sklearn.ClassDictUtil.checkSize(ClassDictUtil.java:130)

The Java implementation of `Hasher` is probably reporting its "output schema" (eg. the number of columns produced) incorrectly.

If you're looking to "debug" the unpickled object,  then you should be using the utility method `ClassDictUtil#toString(ClassDict)` for that. However, it is important that you invoke it before doing any real work in the Java side:

```Java
public class Hasher extends Transformer {

  @Override
  public List<Feature> encodeFeatures(List<Feature> features, SkLearnEncoder encoder){
    System.out.println(ClassDictUtil.toString(this)); // THIS!
    // Proceed as usual
  }
}
```

**Comment by nejatb** [comment]
Hi Villu,

Thank you for the debug suggestion. It was super helpful in understanding what values the class holds. Yes, you are right. The python code Hasher accepts a column of strings and produces a column of numbers by taking their hash. The transform method in python looks like this:

     def transform(self, raw_X):
        n_samples = len(raw_X)
        X = np.zeros((n_samples,self.n_features))
        for i in range(0, ):
            X[i][0] = float(murmurhash(raw_X[i]))
            self.FHvocabulary[raw_X[i]] = float(murmurhash(raw_X[i]))
        return X

and my java side looks like:

     @Override
        public List<Feature> encodeFeatures(List<Feature> inputFeatures, SkLearnEncoder encoder) {

        System.out.println(ClassDictUtil.toString(this));
        Map<String, ?> features = getVocabulary();
        Set<String> featureNames = features.keySet();
        List<Feature> result = new ArrayList<>();

        ClassDictUtil.checkSize( 1, inputFeatures);
        Feature feature = inputFeatures.get(0);

        for (String term: featureNames){
            System.out.println(term);
            Feature feature1 = new Feature(encoder, FieldName.create( "Feature_Hasher(" + feature.getName().getValue() + ")"), DataType.DOUBLE) {
                @Override
                public ContinuousFeature toContinuousFeature() {
                    DerivedField derivedField = createHasherImplementationInPMML(term);
                    return new ContinuousFeature(encoder, derivedField);
                }

            };
            result.add(feature1);
           }

        return result;
     }


The reason I'm getting the `java.lang.IllegalArgumentException: Expected 64 element(s), got 66 element(s)` is probably because I'm looping over all terms in the vocabulary key set and in my training set, there were only 3 distinct values for Feature3. But I looked into the Feature class and I couldn't find the feature's actual value in it. For example, given the mapper above, in the Feature passed to encodeFeatures, I see the following:
`feature.getName()` returns `feature3`
`feature.ref().getField().getValue()` also returns `feature3`
and I couldnt find any other variable that would hold the actual value of the feature, which may be **"abc"**. Am I looking in the wrong place for the value of the feature?

**Comment by vruusmann** [comment]
These are the two key ideas that you need to keep in mind when designing PMML encoders:
1. PMML is a "symbolic" representation. Basically, you'll be defining input and output fields ("variables"), and expressing the logic that connects them.
2. PMML is a "one-dimensional" representation. You'll be operating on a single data record ("row") at a time.

Based on the above ideas, the PMML representation of your `Hasher` transformation would look like this:
```XML
<DerivedField name="hasher(fieature3)" optype="continuous" dataType="double">
  <Apply function="x-hash">
    <FieldRef field="feature3"/>
    <Constant>murmur3_32</Constant>
  </Apply>
</DerivedField>
```

For reference, the Java equivalent of the above would be this:
```Java
String feature3 = ...
double hashedFeature3 = x_hash(feature3, "murmur3_32")
```

I'm defining a generic string hashing function `x-hash` here. It takes two arguments, the string value to be hashed, and the name of the hashing algorithm (since you're trying to reproduce Scikit-Learn behaviour, then it needs to be "signed 32-bit version of Murmurhash3").

**Comment by vruusmann** [comment]
Actually, if you're dealing with a fixed vocabulary, and don't want to bother yourself with (re-)implementing the `x-hash` function on the Java side, then it might be possible to encode your `Hasher` transformation as a simple look-up table - the key column is the string input value, and the value column is the corresponding "signed 32-bit version of Murmurhash3" output value.

This mapping would be represented using the `MapValues` element:
http://dmg.org/pmml/v4-3/Transformations.html#xsdElement_MapValues

There's a fair chance that you can use the newly introduced `sklearn2pmml.preprocessing.LookupTransformer` for this:
https://github.com/jpmml/sklearn2pmml/issues/70

Creating and using the lookup table:
```Python
string2hash = dict()
for word in vocabulary:
  string2hash[word] = float(murmurhash(word))

mapper = DataFrameMapper([
  ("feature3", [CategoricalDomain(), LookupTransformer(string2hash, 0.0)]) # If the word is not part of the "supported vocabulary", then return 0.0
])
```

This would give you 100% pure PMML solution, whereas when going down the earlier route you'd be dealing with a non-standard `x-hash` function. I can imagine/speculate that a dictionary lookup would also be faster than computing the hash "from scratch" every time.

**Comment by nejatb** [comment]
Thanks Villu. Your explanation helped a lot. What I also learned was that for each transformer added, one needs to add a DefineFunction as well that describes the logic of the transformer. Is that right?
For my case, the define function looked similar to this, where I set the offset to 0 to preserve the hashed value:

    <DefineFunction name="feature_hasher" optype="continuous" dataType="double">
			<ParameterField name="hashed_val"/>
			<ParameterField name="murmur3_32"/>
			<ParameterField name="offset"/>
			<Apply function="+">
				<FieldRef field="hashed_val"/>
				<FieldRef field="offset"/>
			</Apply>
    </DefineFunction>

I had one question:
The apply function, does it have to be one of the built-in ones described here (http://dmg.org/pmml/v4-3/BuiltinFunctions.html#uppercase)? or can one potentially create a new apply function as well?

**Comment by vruusmann** [comment]
> One needs to add a DefineFunction as well that describes the logic of the transformer. Is that right?

In most cases it's possible to implement transformation logic in the body of a `DerivedField` element.

You would need to introduce a `DefineFunction` element if there's an opportunity to **reuse** a non-trivial amount of transformation logic between different `DerivedField` elements. It's the same principle as with everywhere else in software engineering - if the same piece of application logic is needed in different places, then you'd extract it into an utility method.

Having a `DefinedFunction` element around is an extra cost (both for model development/conversion, and model execution perspective). So you should also make sure that the benefits are outweighing the costs. For example, it doesn't make much sense to create a `DefineFunction` element that contains a single `Apply` element. It would be much simpler if this `Apply` element was "inlined" where it is actually needed.

If you take a look into Scikit-Learn exported TF-IDF models, then the `DefineFunction` element is only used to encapsulate the `TextIndex` expression (as it can be quite large, especially if there is text normalization involved). All other expressions are "inlined".

> For my case, the define function looked similar to this, where I set the offset to 0 to preserve the hashed value

During conversion, the aim is to "prune" transformation logic from everything unnecessary. In your case, if you know that the value of the `offset` parameter is always 0, then you should omit this `hashed_val + offset` expression altogether. The same goes for the `murmur3_32` parameter - it doesn't seem to be used for anything.

> The apply function, does it have to be one of the built-in ones, or can one potentially create a new apply function as well?

The `Apply@function` attribute refers to some function by name:
1. PMML built-in functions.
2. PMML-backed user-defined functions - the ones that are defined by `DefineFunction` elements.
3. Java-backed user-defined functions. This is vendor specific. For example, the JPMML-Evaluator library lets you invoke any transformation logic that is available as a 3rd party Java library.

If your requirement is to incorporate some exotic string hashing function into the workflow, then the third approach (implementing it as a Java-backed UDF) is the that I would recommend.

**Comment by nejatb** [comment]
Thanks. I need to use a java-backed user-defined function. Is there any example of how to incorporate a java-backed UDF in the transformer?

**Comment by vruusmann** [comment]
> Is there any example of how to incorporate a java-backed UDF in the transformer?

The PMML side:
```XML
<Apply function="com.mycompany.SomeFunction">
  <!-- function arguments go here -->
</Apply>
```

The Java side:
```Java
package com.mycompany;

public class SomeFunction extends org.jpmml.evaluator.functions.AbstractFunction {

  @Override
  public FieldValue evaluate(List<FieldValue> arguments){
    // Java-backed business logic goes here
  }
}
```

Put the `SomeFunction` class into a shareable JAR file, and append it to your Java application classpath.

The use of Java-backed UDFs has been discussed and exemplified earlier in GitHub and JPMML mailing list. Most of those principles/techniques hold true to this day, although there are some new "productivity hacks" available.

**Comment by nejatb** [comment]
Thank you so much. Your guidance was super helpful. I will include my java and PMML side here for reference and will close this issue.

Java side:
    
```java
    public class Hasher extends Transformer{

    public Hasher(String module, String name) {
        super(module, name);
    }

    @Override
    public OpType getOpType(){
        return OpType.CATEGORICAL;
    }

    @Override
    public DataType getDataType(){
        return DataType.STRING;
    }

    @Override
    public List<Feature> encodeFeatures(List<Feature> features, SkLearnEncoder encoder) {
        ClassDictUtil.checkSize(1,features); // Consume 1-D input
        Feature feature = features.get(0);
        Apply apply = encodeApply("com.mycompany.MurmurHash3",feature,"feature_3");
        DerivedField derivedField = encoder.createDerivedField(FieldName.create("feature_hasher"),OpType.CATEGORICAL,DataType.DOUBLE,apply);
        return Collections.singletonList(new ContinuousFeature(encoder, derivedField)); // Produce 1-D output
    }

    public Apply encodeApply(String function, Feature feature, String term){
        Constant constant = PMMLUtil.createConstant(term)
                                .setDataType(DataType.STRING);
        Apply apply = PMMLUtil.createApply(function, feature.ref(), constant);
        return apply;
    }
    }
```

PMML side:

```
<DerivedField name="feature_hasher" optype="categorical" dataType="double">
	<Apply function="com.mycompany.MurmurHash3">
		<FieldRef field="feature_3"/>
		<Constant dataType="string">feature_3</Constant>
	</Apply>
</DerivedField>
```
